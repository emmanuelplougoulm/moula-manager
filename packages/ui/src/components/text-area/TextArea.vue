<template>
  <div class="container">
    <LabelItem
      :text="label"
      :highlighted="isHighlighted"
      :tooltip-title="props.tooltipTitle"
      :tooltip-description="props.tooltipDescription"
    />
    <textarea
      id="text-area"
      data-testid="text-area"
      type="text"
      :placeholder="placeholder"
      :class="textAreaClass"
      class="text-big-regular text-area--default"
      :value="modelValue"
      @input="onChange"
    />
    <ErrorItem :text="errorText" />
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import ErrorItem from '../error/ErrorItem.vue';
import LabelItem from '../label/LabelItem.vue';

import { getErrorText } from '../../utils/commons';

const props = defineProps({
  label: {
    type: String,
    required: true
  },
  placeholder: {
    type: String,
    default: ''
  },
  modelValue: {
    type: String,
    required: true
  },
  pattern: {
    type: RegExp,
    required: true
  },
  required: {
    type: Boolean,
    required: true
  },
  tooltipTitle: {
    type: String,
    default: undefined
  },
  tooltipDescription: {
    type: String,
    default: undefined
  }
});

// eslint-disable-next-line vue/no-setup-props-destructure
const pattern = props.pattern;

let errorText = '';
let textAreaClass = props.required ? 'text-area--highlighted' : '';

const emit = defineEmits(['update:modelValue']);

const isHighlighted = ref(props.required);
const isValid = ref(pattern.test(props.modelValue));

function onChange(event: any): void {
  isValid.value = pattern.test(event.target.value);

  if (props.required) {
    isHighlighted.value = !isValid.value;

    if (isHighlighted.value) {
      errorText = getErrorText(props.required, event.target.value);
      textAreaClass = 'text-area--highlighted';
    } else {
      errorText = '';
      textAreaClass = '';
    }
  }

  if (!props.required) {
    if (!isValid.value) {
      errorText = getErrorText(props.required, event.target.value);
    } else {
      errorText = '';
    }
  }
  emit('update:modelValue', event.target.value);
}
</script>

<style scoped>
.container {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

textarea {
  resize: auto;
  padding: 0.75rem 1rem;
  min-height: 7.5rem;
}

textarea:focus {
  outline: 0;
  background-color: var(--neutral-70);
  border: none;
  overflow: hidden;
  box-shadow: none;
}

.text-area--default {
  border-radius: 0.25rem;
  border: none;
  background-color: var(--neutral-80);
  color: var(--neutral-30);
  box-sizing: border-box;
}

.text-area--highlighted {
  border-radius: 0.25rem;
  border: 0.125rem var(--primary-30) solid;
  background-color: var(--primary-90);
  color: var(--primary-30);
}
</style>
