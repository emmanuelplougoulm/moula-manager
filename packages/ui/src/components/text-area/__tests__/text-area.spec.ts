import { mount } from '@vue/test-utils';
import TextArea from '../TextArea.vue';

describe('TextArea tests', () => {
  it('should display highlighted class on label and text-area', () => {
    const wrapper = mount(TextArea, {
      props: {
        label: 'Description',
        required: true,
        pattern: /^([ \w]{1,128})(?![^?!§&#{[@\]}=+])/,
        modelValue: ''
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display default class on label and focus pseudo class on text-area', async () => {
    const wrapper = mount(TextArea, {
      props: {
        label: 'Informations',
        required: true,
        pattern: /^([ \w]{1,128})(?![^?!§&#{[@\]}=+])/,
        modelValue: ''
      }
    });

    const textArea = wrapper.find('textarea[type="text"]');
    await textArea.setValue('2, place Citron');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display highlighted class on label text-area due to invalid format', async () => {
    const wrapper = mount(TextArea, {
      props: {
        label: 'Adresse complète',
        required: true,
        pattern: /^([ \w]{1,128})(?![^?!§&#{[@\]}=+])/,
        modelValue: ''
      }
    });

    const textArea = wrapper.find('textarea[type="text"]');
    await textArea.setValue('*ù');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display default class on label and text-area', async () => {
    const wrapper = mount(TextArea, {
      props: {
        label: 'Adresse complète',
        required: false,
        pattern: /^([ \w]{1,128})(?![^?!§&#{[@\]}=+])/,
        modelValue: ''
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
