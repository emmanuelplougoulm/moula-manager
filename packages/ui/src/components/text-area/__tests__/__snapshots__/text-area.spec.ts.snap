// Vitest Snapshot v1

exports[`TextArea tests > should display default class on label and focus pseudo class on text-area 1`] = `
"<div data-v-370ce806=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-370ce806=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Informations <!--v-if--></span><textarea data-v-370ce806=\\"\\" id=\\"text-area\\" data-testid=\\"text-area\\" type=\\"text\\" placeholder=\\"\\" class=\\"text-area--highlighted text-big-regular text-area--default\\"></textarea>
  <!--v-if-->
</div>"
`;

exports[`TextArea tests > should display default class on label and text-area 1`] = `
"<div data-v-370ce806=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-370ce806=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default\\" data-testid=\\"label-item\\">Adresse complète <!--v-if--></span><textarea data-v-370ce806=\\"\\" id=\\"text-area\\" data-testid=\\"text-area\\" type=\\"text\\" placeholder=\\"\\" class=\\"text-big-regular text-area--default\\"></textarea>
  <!--v-if-->
</div>"
`;

exports[`TextArea tests > should display highlighted class on label and text-area 1`] = `
"<div data-v-370ce806=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-370ce806=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Description <!--v-if--></span><textarea data-v-370ce806=\\"\\" id=\\"text-area\\" data-testid=\\"text-area\\" type=\\"text\\" placeholder=\\"\\" class=\\"text-area--highlighted text-big-regular text-area--default\\"></textarea>
  <!--v-if-->
</div>"
`;

exports[`TextArea tests > should display highlighted class on label text-area due to invalid format 1`] = `
"<div data-v-370ce806=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-370ce806=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Adresse complète <!--v-if--></span><textarea data-v-370ce806=\\"\\" id=\\"text-area\\" data-testid=\\"text-area\\" type=\\"text\\" placeholder=\\"\\" class=\\"text-area--highlighted text-big-regular text-area--default\\"></textarea>
  <!--v-if-->
</div>"
`;
